Intersection only predicates

thomcom opened the pull request:
## Description
Closes #1015
Depends on #1009 

This PR implements `intersects` and all of the feature combinations that depend exclusively on intersects, as listed in #1015.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

isVoid commented on python/cuspatial/cuspatial/core/binpreds/feature_intersects.py:
```diff
@@ -33,22 +111,132 @@ def _preprocess(self, lhs, rhs):
+        x_coords = rhs.lines.x
+        y_coords = rhs.lines.y
+        result = _false_series(len(lhs))
+        for idx in range(len(intersections)):
+            if isinstance(intersections[idx], ShapelyLineString):
+                result[match_indices[idx]] = True
+                continue
+            x_match = intersections.points.x[idx] == x_coords
+            if x_match.sum() > 0:
+                y_match = intersections.points.y[idx] == y_coords
+                if (y_match & x_match).sum() > 0:
+                    result[match_indices[idx]] = True
+        return result
```
Hmm....

I'd suggest can we leave out this specific combination for this release, and wait until we have #1022 ?

thomcom replied:
Yes, I agree. I actually think I worked this to the point of implementing `allpairs_multipoint_equals_count` and then branched and forgot I'd added this.

thomcom replied:
Indeed there's still a failing test from working on this feature directly before pairing with you. :D

isVoid commented on python/cuspatial/cuspatial/core/binpreds/feature_intersects.py:
```diff
@@ -33,22 +111,132 @@ def _preprocess(self, lhs, rhs):
+
+class PolygonLineStringIntersects(IntersectsPredicateBase):
+    def _preprocess(self, lhs, rhs):
+        """Convert rhs to linestrings."""
```
```suggestion
        """Convert lhs to linestrings."""
```

thomcom replied:
done

isVoid commented on python/cuspatial/cuspatial/core/binpreds/feature_intersects.py:
```diff
@@ -33,22 +111,132 @@ def _preprocess(self, lhs, rhs):
+        return result
+
+
+class LineStringPolygonIntersects(IntersectsPredicateBase):
```
Polygon related intersection tests should involve some kind of Point-In-Polygon test... The linestring and the linestrings made from the rings can have 0 intersection, but the linestring can still have intersection with the polygon (contained).

thomcom replied:
I see your point, I had incorrectly assumed this behavior. I'll look at integrating intersects and pip for this now.

thomcom replied:
Made into `NotImplementedPredicates`.

isVoid reviewed (changes requested):
Mostly good, two things sticks out:
1. Polygon intersection seems a bit flawed.
2. Multipoint-linestring intersection has a host for loop on the result of intersections, quite not ideal...

thomcom commented on python/cuspatial/cuspatial/tests/binpreds/test_intersects_only_binpreds.py:
```diff
@@ -0,0 +1,982 @@
+    pd.testing.assert_series_equal(expected, got.to_pandas())
+
+
+def test_linestring_intersects_polygon():
```
Indeed, there are many special cases of polygons that will need to have tests written for them. I'm planning on writing a suite of special case tests when the dispatched tests, in the next upcoming PR for 23.06, are complete. All of the special case tests will be documented and need pretty careful scrutiny.

How do you feel about postponing them until then so that this can get into 23.04? Exhaustively covering special cases is going to be a multi-day effort I think.

isVoid replied:
> How do you feel about postponing them until then so that this can get into 23.04?

I'm totally fine with postponing as long as they are tracked. You can convert this comment into an issue, for example.

thomcom replied:
Thank you, and [done](https://github.com/rapidsai/cuspatial/issues/1036)

thomcom commented on python/cuspatial/cuspatial/core/binops/intersection.py:
```diff
@@ -8,15 +10,17 @@
     empty_geometry_column,
 )
 
+if TYPE_CHECKING:
+    from cuspatial.core.geoseries import GeoSeries
+
```
I was under the impression that I still needed it, but upon testing without it see that you are right. Thanks!

thomcom commented on python/cuspatial/cuspatial/core/binops/intersection.py:
```diff
@@ -97,6 +103,8 @@ def pairwise_linestring_intersection(
     meta = GeoMeta(
         {"input_types": types_buffer, "union_offsets": offset_buffer}
     )
+    from cuspatial.core.geoseries import GeoSeries
```
Ah, yes, I must have inserted the first one and then blindly inserted the second one assuming the scopes were different.

thomcom commented on python/cuspatial/cuspatial/core/binpreds/feature_contains.py:
```diff
@@ -356,7 +355,7 @@ def _preprocess(self, lhs, rhs):
     (MultiPoint, LineString): NotImplementedPredicate,
     (MultiPoint, Polygon): NotImplementedPredicate,
-    (LineString, Point): NotImplementedPredicate,
+    (LineString, Point): PointPointContains,
```
Yes, they both simply return the result of `x.intersects(y)`. So far I haven't renamed any dispatched predicates to something generic until (if) they are imported by another file.

thomcom commented on python/cuspatial/cuspatial/core/binpreds/feature_disjoint.py:
```diff
@@ -0,0 +1,70 @@
+
+class ContainsDisjoint(BinPred):
+    def _preprocess(self, lhs, rhs):
+        """Disjoint is the opposite of contains, so just implement contains
+        and then negate the result."""
```
That's a good idea. I probably have time for this.

harrism replied:
How is disjoint the opposite of contains for polygons? A polygon that is NOT disjoint from another polygon or a linestring does not necessarily contain it -- for example if they intersect.  Intersecting geometries are not disjoint and don't contain each other, except where the intersection is only a colinear point or segment.

thomcom replied:
Thanks for catching this as well. Disjoint is the opposite of contains for (Point, Point), (Point, Polygon), (Polygon, Point), and (Polygon, MultiPoint). I mistakenly rushed into applying the same reversal to (Polygon, LineString) and (Polygon, LineString), possibly because I trusted my own incorrect docstring too much. I will modify the latter two types to `NotImplementedPredicate` and correct the docstring to better reflect the correct reversal.

As I write the remaining tests in the test_dispatch (next PR for 23.06) I will first make a list of the unimplemented cases (including the latter two here) and then cross-reference them all with the [Binary Predicate Design Doc](https://docs.google.com/document/d/19BzZ-C0e4vBl9tAMJeDUc9s6AhhAs8cYOGXkWCawbLY/edit?pli=1#heading=h.d8e4dm4w5mzd) rules. I'll use that for the rest of the more complicated relationships so I don't stumble down as many rabbit holes.

thomcom replied:
[Enumerate non-simple types left to implement](https://github.com/rapidsai/cuspatial/issues/1037)

thomcom commented on python/cuspatial/cuspatial/core/binpreds/feature_intersects.py:
```diff
@@ -1,27 +1,105 @@
+        basic_result = pairwise_linestring_intersection(
+            preprocessor_result.lhs, preprocessor_result.rhs
+        )
+        computed_result = IntersectsOpResult(basic_result)
```
Done. This is why I have objects for connecting the pipeline stages, I wish I could enforce even stronger type checking here especially.

isVoid replied:
Would you like to? We can explore adding type checks in 23.06.

thomcom commented on python/cuspatial/cuspatial/core/binpreds/feature_intersects.py:
```diff
@@ -1,27 +1,105 @@
+        correct type for the predicate."""
+        match_indices = self._get_match_indices(lhs, op_result)
+        result = _false_series(len(lhs))
+        if len(op_result.result[1]) > 0 and len(lhs) == 1:
+            result[0] = True
```
Another good find. The first case was an intermediate behavior, maybe before I had `_get_intersecting_geometry_indices` working fully, and the second case is the proper handling of all existing cases. Simplified.

thomcom commented on python/cuspatial/cuspatial/core/binpreds/feature_intersects.py:
```diff
@@ -33,22 +111,132 @@ def _preprocess(self, lhs, rhs):
 
+class LineStringPointIntersects(IntersectsPredicateBase):
+    def _preprocess(self, lhs, rhs):
+        """Convert rhs to linestrings."""
```
Good idea done.

thomcom commented:
I handled Michael's comments, which mostly had to do with trying to incorrectly implement predicates that depend on

1. An unimplemented feature (`allpairs_multipoint_equals_count`)
2. Predicates that depend on combining the results of `intersects` and `contains_properly`.

The scope of this PR is not to implement either of these cases yet, so I replaced the offending predicates with `NotImplementedPredicate` and cleaned things up otherwise. Please reconsider this submission for 23.04.

isVoid reviewed (approved):
The file name for binpreds/feature_* is a bit redundant, calling them with just the name of the predicate would be sufficient.

Else all is good. Great work!

thomcom commented:
Part of the reason for naming them as such is so that they'll be grouped together. Without the naming scheme then as more files are created it won't be as clear from looking at the file list how they can be grouped. From your approval I'll assume it is ok to leave it as is for now?

isVoid commented:
> From your approval I'll assume it is ok to leave it as is for now?

Yes, file name concerns should not block merging this PR.

harrism commented:
> Part of the reason for naming them as such is so that they'll be grouped together. Without the naming scheme then as more files are created it won't be as clear from looking at the file list how they can be grouped. From your approval I'll assume it is ok to leave it as is for now?

Put them in a directory with the common part of the name?

harrism commented on python/cuspatial/cuspatial/core/binpreds/feature_overlaps.py:
```diff
@@ -20,17 +23,18 @@ class OverlapsPredicateBase(EqualsPredicateBase):
     equals predicate for all implementations up to this point.
```
"up to this point"? What does this mean?

thomcom replied:
All implementations at this point in time. Will clarify.

harrism commented on python/cuspatial/cuspatial/core/binpreds/feature_overlaps.py:
```diff
@@ -20,17 +23,18 @@ class OverlapsPredicateBase(EqualsPredicateBase):
     equals predicate for all implementations up to this point.
     For example, a Point-Point Crosses predicate is defined in terms
```
Crosses? I thought this was Overlaps.

thomcom replied:
typo, fixed

harrism commented on python/cuspatial/cuspatial/core/binpreds/feature_overlaps.py:
```diff
@@ -20,17 +23,18 @@ class OverlapsPredicateBase(EqualsPredicateBase):
     equals predicate for all implementations up to this point.
     For example, a Point-Point Crosses predicate is defined in terms
```
Below it says points can't overlap points.

thomcom replied:
That is a removed block, previously I was creating the `PointPointOverlaps` predicate class, but now I just need `ImpossiblePredicate` to return `False` for all the pairs.

harrism commented on python/cuspatial/cuspatial/core/binpreds/feature_touches.py:
```diff
@@ -0,0 +1,58 @@
+
+
+DispatchDict = {
+    (Point, Point): PointPointTouches,
```
If points can't touch, shouldn't you dispatch to Impossible instead of a custom implementation?

thomcom replied:
Yes, thank you.

harrism commented on python/cuspatial/cuspatial/core/geoseries.py:
```diff
@@ -1226,3 +1215,26 @@ def crosses(self, other, align=True):
+        Returns
+        -------
+        result : cudf.Series
+            A Series of boolean values indicating whether each geometry
```
```suggestion
            A Series of Boolean values indicating whether each pair of corresponding geometries is disjoint.
```

thomcom replied:
Done.

harrism commented on python/cuspatial/cuspatial/tests/binpreds/test_intersects_only_binpreds.py:
```diff
@@ -0,0 +1,1017 @@
+    pd.testing.assert_series_equal(expected, got.to_pandas())
+
+
+def test_extra_linestring_intersects_point():
```
Why are there tests called "extra" and "extra 2", etc? Why not just have multiple test cases? Or just number them? Or more descriptive names?

thomcom replied:
Oh just because the descriptive names can get very long. I could just number these, this pair of tests just tests that the beginning point of the linestring results in correct results, and also that the ending point of the linestring does.

harrism reviewed (changes requested):
There is a lot of inconsistency between docs and dispatch.

I feel like a complex combinatoric problem like this should have an implementation that follows a detailed specification. You made a good start at writing a specification, but I think that spec needs more review and revision and then the implementation should be closely tied to it.

thomcom commented on python/cuspatial/cuspatial/core/binpreds/feature_contains.py:
```diff
@@ -356,7 +360,7 @@ def _preprocess(self, lhs, rhs):
     (MultiPoint, LineString): NotImplementedPredicate,
     (MultiPoint, Polygon): NotImplementedPredicate,
-    (LineString, Point): NotImplementedPredicate,
+    (LineString, Point): PointPointContains,
```
In this case, LineStringPoint reuses the logic of PointPointContains. I've renamed it to `ContainsByIntersection`.

thomcom commented on python/cuspatial/cuspatial/core/binpreds/feature_crosses.py:
```diff
@@ -35,7 +45,7 @@ def _preprocess(self, lhs, rhs):
     (MultiPoint, LineString): NotImplementedPredicate,
     (MultiPoint, Polygon): NotImplementedPredicate,
-    (LineString, Point): NotImplementedPredicate,
+    (LineString, Point): ImpossiblePredicate,
```
Because GeoPandas returns False for the combinations of LineString, Point:
![lines-dont-cross-points](https://user-images.githubusercontent.com/410931/230460474-4c5c0dc4-5d47-4d8c-9920-857e579ce027.png)

thomcom commented on python/cuspatial/cuspatial/core/geoseries.py:
```diff
@@ -1226,3 +1215,26 @@ def crosses(self, other, align=True):
+        """Returns True for all aligned geometries that are disjoint from
+        other, else False.
+
+        Geometries are disjoint if they do not share any points in common.
```
I updated this with better text "An object is said to be disjoint to other if its boundary and interior does not intersect at all with those of the other".

thomcom commented:
Resolved @harrism's comments and merging for 23.04.

thomcom commented:
/merge
